BUG: f2py: better handle filtering of public/private subroutines

m-weigand opened the pull request:
This should fix gh-26920 and add a corresponding test case.

I am unsure about some formalities and would appreciate some feedback:

* I am not sure if I put the test in the correct location
* Technically, this fixes another underlying bug: Handling of f90 modules that have no public variables or subroutines. I'm not sure if this warrants a separate test, but please let me know and I will reformat accordingly.

m-weigand committed:
TST:  Add regression test for gh-26920

m-weigand committed:
BUG: f2py: better handle filtering of public/private subroutines

Don't mistake public/private declarations of F90 subroutines for variables
when the corresponding subroutines are filtered by use of only:.
Also, handle modules with no public variables or subroutines, caused by
the filtering.

Closes gh-26920.

HaoZeke commented on numpy/f2py/f90mod_rules.py:
```diff
@@ -110,11 +110,22 @@ def dadd(line, s=doc):
-            if (n not in notvars) and (not l_or(isintent_hide, isprivate)(var)):
+            # heuristic to find public/private declarations of filtered
+            # subroutines
+            if len(var) == 1 and 'attrspec' in var and var['attrspec'][0] in ('public', 'private'):
+                is_not_var = True
```
I wonder if this entire check could be extracted into a functional form in `auxfuncs.py` instead. Similar to the existing:

```python
def isprivate(var):
    return 'attrspec' in var and 'private' in var['attrspec']
```

This is a bit nitpicky and mostly for consistency with the rest of the F2PY code, the form here also works.

HaoZeke replied:
Also, what about the `protected` access property? (not sure how / if this is handled in F2PY now actually).

m-weigand replied:
As far as I can see, the `protected` property is correctly handled, i.e., using it produced no errors for me. However, I'm not familiar enough with it to tell if I missed anything here.

HaoZeke reviewed (commented):
I have two minor comments, and beyond that I think it would be nice to have a separate test for the no public subroutine or variable case, otherwise, this looks great.

EwoutH commented:
It this the last PR that needs to be merged for a 2.1.0 release candidate?

charris commented:
> It this the last PR that needs to be merged for a 2.1.0 release candidate?

Not for the rc, and maybe not for the final either if it isn't resolved.

m-weigand commented:
Sorry for the late reply, summer vacation got in the way.

I modified the pull request to include:

* an addition example with no public entities in both modules
* I moved the check in **f90mod_rules.py** into **auxfuncs.py**

Please let me know if I should change anything else.

HaoZeke commented:
LGTM, in it goes, thanks for working on this @m-weigand !
